util/EFSOI_Utilities/src/statevec_efsoi.f90

```diff
@@ -0,0 +1,326 @@
+call gettable(tbname,luin,ntot,nvars,utable)
+
+! release file unit
+close(luin)
+
+! Retrieve each token of interest from table and define
+! variables participating in control vector
+
+! Count variables first
+nc2d=0; nc3d=0; ncdim=0;
+do ii=1,nvars
+   read(utable(ii),*) var, ilev, itracer, source, funcof
+   if(ilev==1) then
+       nc2d=nc2d+1
+       ncdim=ncdim+1
+   else
+       nc3d=nc3d+1
+       ncdim=ncdim+ilev
+   endif
+enddo
+
+allocate(cvars3d(nc3d),cvars2d(nc2d),clevels(0:nc3d))
+
+! Now load information from table
+nc2d=0;nc3d=0
+clevels = 0
+do ii=1,nvars
+   read(utable(ii),*) var, ilev, itracer, source, funcof
+   if(ilev==1) then
+      nc2d=nc2d+1
+      cvars2d(nc2d)=trim(adjustl(var))
+   else if (ilev==nlevs .or. ilev==nlevs+1) then
+      nc3d=nc3d+1
+      cvars3d(nc3d) = trim(adjustl(var))
+      clevels(nc3d) = ilev + clevels(nc3d-1)
+   else
+      if (nproc .eq. 0) print *,'Error: only ', nlevs, ' and ', nlevs+1,' number of levels is supported in current version, got ',ilev
+      call stop2(503)
+   endif
+enddo
+
+deallocate(utable)
+
+allocate(index_pres(ncdim))
+ii=0
+do i=1,nc3d
+  do k=1,clevels(i)-clevels(i-1)
+    ii = ii + 1
+    index_pres(ii)=k
+  end do
+end do
+do i = 1,nc2d
+  ii = ii + 1
+  index_pres(ii) = nlevs+1
+enddo
+
+! sanity checks
+if (ncdim == 0) then
+  if (nproc == 0) print *, 'Error: there are no variables to update.'
+  call stop2(501)
+endif
+
+do i = 1, nc2d
+  if (getindex(vars2d_supported, cvars2d(i))<0) then
+    if (nproc .eq. 0) then
+      print *,'Error: 2D variable ', cvars2d(i), ' is not supported in current version.'
+      print *,'Supported variables: ', vars2d_supported
+    endif
+    call stop2(502)
+  endif
+enddo
+do i = 1, nc3d
+  if (getindex(vars3d_supported, cvars3d(i))<0) then
+    if (nproc .eq. 0) then 
+       print *,'Error: 3D variable ', cvars3d(i), ' is not supported in current version.'
+       print *,'Supported variables: ', vars3d_supported
+    endif
+    call stop2(502)
+  endif
+enddo
+
+if (nproc == 0) then 
+  print *, '2D control variables: ', cvars2d
+  print *, '3D control variables: ', cvars3d
+  print *, 'Control levels: ', clevels
+  print *, 'nc2d: ', nc2d, ', nc3d: ', nc3d, ', ncdim: ', ncdim
+endif
+
+
+call getgridinfo(fgfileprefixes(1), reducedgrid)
+
+
+! Identify EFSOI relevant state variable indices
+u_ind   = getindex(vars3d_supported, 'u')   !< indices in the state var arrays
+v_ind   = getindex(vars3d_supported, 'v')   ! U and V (3D)
+tv_ind  = getindex(vars3d_supported, 'tv')  ! Tv (3D)
+q_ind   = getindex(vars3d_supported, 'q')   ! Q (3D)
+ps_ind  = getindex(vars2d_supported, 'ps')  ! Ps (2D)
+
+! Index of each elements
+allocate(id_u(nlevs))
+allocate(id_v(nlevs))
+allocate(id_t(nlevs))
+allocate(id_q(nlevs))
+do k=1,nlevs
+   id_u(k) = clevels(u_ind-1) + k
+   id_v(k) = clevels(v_ind-1) + k
+   id_t(k) = clevels(tv_ind-1) + k
+   id_q(k) = clevels(q_ind-1) + k
+end do
+id_ps = clevels(nc3d) + ps_ind
+
+
+
```

Comment:
@AndrewEichmann-NOAA Another set of blank lines

Reply:
removed